src/equihash.rs

```diff
@@ -0,0 +1,272 @@
+use blake2b::{Blake2b, Hash, ParameterBlock};
+use byteorder::{LittleEndian, WriteBytesExt};
+
+struct Params {
+    n: u32,
+    k: u32,
+}
+
+#[derive(Clone)]
+struct Node {
+    hash: Vec<u8>,
+    indices: Vec<u32>,
+}
+
+impl Params {
+    fn indices_per_hash_output(&self) -> u32 {
+        512 / self.n
+    }
+    fn hash_output(&self) -> u8 {
+        (self.indices_per_hash_output() * self.n / 8) as u8
+    }
+    fn collision_bit_length(&self) -> usize {
+        (self.n / (self.k + 1)) as usize
+    }
+    fn collision_byte_length(&self) -> usize {
+        (self.collision_bit_length() + 7) / 8
+    }
+    fn hash_length(&self) -> usize {
+        ((self.k as usize) + 1) * self.collision_byte_length()
+    }
+}
+
+impl Node {
+    fn new(a: &mut Node, b: &mut Node, trim: usize) -> Self {
+        let hash: Vec<_> = a.hash
+            .iter()
+            .zip(b.hash.iter())
+            .skip(trim)
+            .map(|(a, b)| a ^ b)
+            .collect();
+        let mut indices = Vec::new();
+        if a.indices_before(b) {
+            indices.append(a.indices.as_mut());
```

Comment:
Why not do `.extend()`? Then a and b don't need to be mutable references, and in other parts of your code mutable temporaries don't need to be created.

Reply:
Switched to `.extend()`, and also realised that in the recursive case, I can take ownership of the underlying child Nodes, and extend one's indices with the other to eliminate half of the copying cost.

Reply:
I can't do that in the iterative case because I'm using `chunks()` which is a representation, which I can't take ownership of (unless there's a clever way to do so).